### PR TITLE
[PyTorch] Deferred Initialization via `device='meta'` option

### PR DESCRIPTION
**Pull request by denera**
This is primarily meant to support training very large models in `torch.distributed.fsdp.FullyShardedDataParallel` mode. 

FSDP execution flow requires models to be initialized in serial mode before being passed into an FSDP wrapper that shards their parameters. Without `device='meta'` support, this materializes a copy of every module in memory for every host/device and causes OOM errors for large models. Deferring initialization with `device='meta'` materializes the memory only when the FSDP wrapper recursively executes `module.reset_parameters()` on already sharded modules, thus avoiding unsharded duplicates.

~~Databricks has requested us to accelerate upstreaming this feature for `te.Linear` in support of their immediate work instead of waiting to upstream it until the same feature is implemented for other TransformerEngine modules. I will be filing additional PRs to extend this feature to other modules in the coming days/weeks.~~

The PR also includes an example for demonstrating `device='meta'` use with FSDP. ~~I am planning to convert this into a distributed/FSDP regression test once the meta device feature is implemented for other TE modules.~~

**Edit:** Extending this to all TE modules turned out to be easy enough to roll into this PR instead postponing to a follow-up.

### CODE REVIEW

File: examples/pytorch/fsdp/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Basic Example for Using PyTorch Fully Sharded Data Parallel mode with Transformer Engine
+
```
Comment (ptrendx) [inline examples/pytorch/fsdp/README.md:2]: Add some comment for each of those lines (what does each of the example showcase)?

File: transformer_engine/pytorch/module/linear.py [hunk 2]
```diff
@@ -797,6 +784,63 @@ def __init__(
...
+        is specified, all parameters on a 'meta' device are also materialized on a real cuda
+        device before the values are reset to initial.
+        """
+        if defer_init:
+            return
+
+        # materialize all parameters (move to real device, initialize values)
+        for wname, bname in zip(self.weight_names, self.bias_names):
+            if self._parameters[wname].device == torch.device('meta'):
+                self._parameters[wname].to(device='cuda')
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/linear.py:833]: Even though we are mostly concerned with cuda device here, hardcoding this to cuda does not actually feel correct... Not 100% sure how to change it though

File: examples/pytorch/fsdp/fsdp.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
```
Comment (ptrendx) [inline examples/pytorch/fsdp/fsdp.py:3]: Could you add some comments throughout the scripts telling what you are doing here?

File: transformer_engine/pytorch/module/linear.py (hunk 2 above)
Comment (denera) [reply transformer_engine/pytorch/module/linear.py:827]: Paramater reset is doing a cuda cast here only if the tensor is on a meta device. The intent is that the cuda cast can be circumvented as long as the module is cast to a non-meta device *after* initialization, but before the parameter reset.

For example:
```
model = te.Linear(hidden_size, ffn_hidden_size, device='meta')
model.to(device='cpu')
model = torch.distributed.fsdp.FullyShardedDataParallel(model)
```

Unless I'm mistaken, `model.to(device='cpu')` will not trigger memory allocation because the weight/bias tensors are crated as `torch.empty()` when `device='meta'` during module initialization. Consequently, we get the memory allocation only when FSDP triggers the parameter reset, which will not move the tensor to the cuda device because tensor device is not meta at that time.

I'm not sure if this is necessarily the right programming model here, but I think it might be better than the two other alternatives I've thought of:
- Initialize with `te.Linear(...., device='meta', deferred_device='cpu')`, and have the parameter reset always cast to the deferred device.
- Initialize with `te.Linear(..., device='cpu', defer_init=True)`, and internally use meta device initialization based on the deferred init flag.

Neither of these really mesh with PyTorch native module behavior/API so I thought they were worse than the current implementation in the PR. Happy to change it though if we decide otherwise.

File: transformer_engine/pytorch/module/linear.py [hunk 4]
```diff
@@ -797,6 +784,57 @@ def __init__(
         else:
             self.gemm_bias_unfused_add = False
 
+    def init_weight_tensor(self, weights: torch.Tensor) -> torch.Tensor:
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/linear.py:787]: Should we move these to `base.py` since these are general functions that would be required when we expand this to the other modules?

File: transformer_engine/pytorch/module/linear.py [hunk 5]
```diff
@@ -797,6 +784,57 @@ def __init__(
...
+            self.fp8_meta["update_amax_and_scale_fwd"] = True
+            weights = Float8Tensor(
+                data=weights,
+                fp8_meta=self.fp8_meta,
+                fp8_meta_index=tex.FP8FwdTensors.GEMM1_WEIGHT,
+            )
+
+        return weights
+
+    def reset_parameters(self, defer_init: bool = False) -> None:
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/linear.py:815]: Same as above

File: transformer_engine/pytorch/module/linear.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/module/linear.py:827]: Generally speaking we have the devices hardcoded to cuda for all allocations since that's the only supported execution. This is guarded by checks for all the activations and weights etc. It looks to be the same for this PR.

File: transformer_engine/pytorch/module/linear.py (hunk 4 above)
Comment (ptrendx) [reply transformer_engine/pytorch/module/linear.py:787]: I think we can move them with the follow up PR.

Comment (denera) [reply transformer_engine/pytorch/module/linear.py]: I moved them in this PR and probably need to tinker with it as I test other modules besides Linear in follow-up PRs, but in the meantime please take a look and let me know what you think about the approach.

**Comment by denera** [comment]
@ksivaman @ptrendx Extending this to all TE modules turned out to be easy enough to do in this PR instead of splitting it into multiple. Pushed the update now, monitoring the CI for failures. Please let me know what you think.

File: transformer_engine/pytorch/module/_common.py [hunk 6]
```diff
@@ -187,3 +189,20 @@ def _noop_cat(
...
     # Perform no-op concat
     return _NoopCatFunc.apply(split_ranges, full_tensor, *tensors)
+
+
+@dataclass
+class _ParameterInitMeta:
+    """
+    Stores essential metadata needed to support deferred parameter initialization.
+    """
+    parent: torch.nn.Module
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/_common.py:199]: Why is this required?

File: transformer_engine/pytorch/module/base.py [hunk 7]
```diff
@@ -746,6 +749,49 @@ def get_fp8_weights_empty_tensors(
...
+            fp8_meta_index = self.param_init_meta[name].fp8_meta_index
+            if self.primary_weights_in_fp8 and fp8_meta_index is not None:
+                param = Float8Tensor.to_float8(
+                    param,
+                    fp8_meta=self.fp8_meta,
+                    fp8_meta_index=fp8_meta_index
+                )
+
+            # Redo parameter wrap in case we broke it above
+            setattr(self, name, torch.nn.Parameter(param))
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/base.py:793]: This is required only for the `primary_weights_in_fp8` case, right?

**Review by ksivaman — APPROVED** [review]
LGTM
